File: python/gauss_colmain.py
import time

from matrix_tool import print_matrix, print_vector, append_column, multiply
from check import check_roots
import gauss_klassisch as gk


def solve(A, b):
  S = append_column(A, b)
  for i in xrange(len(S)):
    print("Choosing main for iteration #{}".format(i))
    choose_main(S, i)
    gk.forward(S, i)
  roots = []
  for i in xrange(len(S) - 1, -1, -1):
    roots.append(gk.backward_right(S, i))
  roots.reverse()
  return roots


def swap_rows(A, lhs, rhs):
  row = A[lhs][:]
  A[lhs] = A[rhs]
  A[rhs] = row


def choose_main(A, row):
  max_elem, max_row = A[row][row], row
  for i in xrange(row + 1, len(A)):
    if A[i][row] > max_elem:
      max_elem, max_row = A[i][i], i
  if max_row != row:
    print("Swapping {} and {}".format(row, max_row))
    swap_rows(A, row, max_row)
  else:
    print("No swap, good to go!")


def main():
  from constants import A2_21 as A, b2_21 as b
  start_time = time.time()
  x = solve(A, b)
  solve_time = time.time()
  print("Roots:")
  print_vector(x)
  check_roots(A, x, b)
  det = gk.determinant(A)
  det_time = time.time()
  print
  print("Determinant: {}".format(det))
  print
  inversed = gk.inverse(A)
  inverse_time = time.time()
  print("Inverse matrix:")
  print_matrix(inversed, precision=8)
  print("Check inverse matrix:")
  print_matrix(multiply(inversed, A))

  print("Time elapsed: ")
  print(
      "\tSolution:\t{} ms\n\tDeterminant:\t{} ms\n\tInversion:\t{} ms\n\tOverall:\t{} ms".
      format((solve_time - start_time) * 1000, (det_time - solve_time) * 1000, (
          inverse_time - det_time) * 1000, (inverse_time - start_time) * 1000))


if __name__ == '__main__':
  main()

File: matlab/gauss_colmain.m
A = [
    [188, -32, -9, 7, -61, 49, -53, 23, -31, 62],
    [-32, 101, 11, -32, -80, 1, 53, -7, 94, -109],
    [-9, 11, 231, 96, -62, -12, 72, -40, 72, -77],
    [7, -32, 96, 174, -37, -119, 19, 7, -27, 27],
    [-61, -80, -62, -37, 184, -34, -115, 45, -53, 46],
    [49, 1, -12, -119, -34, 182, 80, -56, -4, 18],
    [-53, 53, 72, 19, -115, 80, 231, -101, 70, -50],
    [23, -7, -40, 7, 45, -56, -101, 227, -10, 23],
    [-31, 94, 72, -27, -53, -4, 70, -10, 241, -155],
    [62, -109, -77, 27, 46, 18, -50, 23, -155, 212],
];

b = [
    [-44],
    [74],
    [87],
    [247],
    [-39],
    [168],
    [192],
    [-199],
    [-194],
    [243],
];

function swap_rows(A, l, r)
  tmp = A(l, :);
  A(l, :) = A(r, :);
  A(r, :) = tmp;
end;

function [B, p] = forward_colmain(A)
  n = size(A, 1);
  B = A;
  p = 0;
  for row = 1:n
    [_, mrow] = max(B(row:n, row));
    mrow = mrow + row - 1
    if mrow ~= row
      disp(sprintf('Swapping %d and %d\n', row, mrow));
      p = p + 1;
      B([row, mrow], :) = B([mrow, row], :);
    end;
    B(row, :) = B(row, :) ./ B(row, row);
    for i = row+1:n
      B(i, :) = B(i, :) - B(row, :) * B(i, row);
    end;
    B
  end;
end;


tic
S = [A b]
[S, p] = forward_colmain(S);
[B, x] = backward_right(S);
B
x
toc

T = triangalize(A)
det = determinant(T) * (-1)^p
